test/e2e/upgrade_cluster_test.go

```diff
@@ -0,0 +1,189 @@
+//go:build e2e
+// +build e2e
+
+package e2e
+
+import (
+	"context"
+	"strings"
+	"sync"
+	"testing"
+
+	"github.com/onsi/gomega"
+	hyperv1 "github.com/openshift/hypershift/api/hypershift/v1beta1"
+	e2eutil "github.com/openshift/hypershift/test/e2e/util"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	nodePoolAnnotationCurrentConfig        = "hypershift.openshift.io/nodePoolCurrentConfig"
+	nodePoolAnnotationCurrentConfigVersion = "hypershift.openshift.io/nodePoolCurrentConfigVersion"
+	hostedClusterUpgradeTestLabel          = "hypershift.openshift.io/upgrade-test"
+)
+
+// TestUpgradeHyperShift validates that a HyperShift Operator upgrade won't unnecessarily cause a node rollout.
+// The test must be triggered via the flag upgrade.run-tests and the HyperShift Operator initial image must be set
+// via the flag e2e.hypershift-operator-initial-image.
+func TestUpgradeHyperShift(t *testing.T) {
+	if globalOpts.RunUpgradeTest {
+		var once sync.Once
+		var cleanup e2eutil.Cleanup
+		var err error
+
+		var mgmtClient crclient.Client
+		var hostedCluster *hyperv1.HostedCluster
+
+		var nodePoolsMap map[string]*hyperv1.NodePool
+
+		hyperShiftOperatorLatestImage := globalOpts.HyperShiftOperatorLatestImage
```

Comment:
Is this manipulating the HO on the management cluster for the test it self?
If so that would impact and invalidate all the other e2e test running on this management.
Unless you are planning to run this in isolation. If so can you please clarify in the PR desc?

Reply:
Yeah the plan is to run this test in isolation in its own ci-operator presubmit. 

Creating two nested HCs using the e2e test framework would result in a very complex code IMHO. 

Will clarify in the PR and test description.

Reply:
Done